core/trino-main/src/main/java/io/trino/execution/scheduler/ScaledWriterScheduler.java

```diff
@@ -93,25 +96,49 @@ private int getNewTaskCount()
             return 1;
         }
 
-        double fullTasks = sourceTasksProvider.get().stream()
-                .filter(task -> !task.getState().isDone())
-                .map(TaskStatus::getOutputBufferStatus)
-                .map(OutputBufferStatus::isOverutilized)
-                .mapToDouble(full -> full ? 1.0 : 0.0)
-                .average().orElse(0.0);
-
         long writtenBytes = writerTasksProvider.get().stream()
                 .map(TaskStatus::getPhysicalWrittenDataSize)
                 .mapToLong(DataSize::toBytes)
                 .sum();
 
-        if ((fullTasks >= 0.5) && (writtenBytes >= (writerMinSizeBytes * maxTaskWriterCount * scheduledNodes.size()))) {
+        // When there is a big data skewness, there could be a bottleneck due to the skewed workers even if most of the workers are not over-utilized.
+        // Check both, weighted output buffer over-utilization rate and average output buffer over-utilization rate, in case when there are many over-utilized small tasks
+        // due to fewer not-over-utilized big skewed tasks.
+        if ((isWeightedBufferUtilizationFull() || isAverageBufferUtilizationFull()) && (writtenBytes >= (writerMinSizeBytes * maxTaskWriterCount * scheduledNodes.size()))) {
             return 1;
         }
 
         return 0;
     }
 
+    private boolean isWeightedBufferUtilizationFull()
+    {
+        double totalOutputSize = 0.0;
+        double overUtilizedOutputSize = 0.0;
+        for (TaskInfo task : sourceTasksProvider.get()) {
+            if (!task.getTaskStatus().getState().isDone()) {
+                long outputDataSize = task.getStats().getOutputDataSize().toBytes();
+                totalOutputSize += outputDataSize;
+                if (task.getTaskStatus().getOutputBufferStatus().isOverutilized()) {
+                    overUtilizedOutputSize += outputDataSize;
+                }
+            }
+        }
+
+        return totalOutputSize > 0 ? overUtilizedOutputSize / totalOutputSize >= BUFFER_UTILIZATION_THRESHOLD : false;
+    }
+
+    private boolean isAverageBufferUtilizationFull()
```

Comment:
I don't think we need this version of check anymore, do we? Whenever `isWeightedBufferUtilizationFull` retruns false, then `isAverageBufferUtilizationFull` would too, right?

Reply:
Take a look at this: https://github.com/trinodb/trino/pull/14874#discussion_r1022219822

Reply:
Can smaller tasks be overutilized while big one isnt?

Reply:
Anyways I don't have strong opinion about it